src/parquet/bloom.cc

```diff
@@ -0,0 +1,217 @@
+    num_bits = static_cast<uint32_t>(::arrow::BitUtil::NextPower2(num_bits));
+  }
+
+  // Minimum
+  if (num_bits < (Bloom::BYTES_PER_FILTER_BLOCK << 3)) {
+    num_bits = Bloom::BYTES_PER_FILTER_BLOCK << 3;
+  }
+
+  return num_bits;
+}
+
+void Bloom::addElement(uint64_t hash) {
+  uint32_t* const bitset32 = reinterpret_cast<uint32_t *>(bitset.get());
+  const uint32_t bucketIndex = static_cast<uint32_t>(hash >> 32)
+      & (num_bytes / BYTES_PER_FILTER_BLOCK - 1);
+  uint32_t key = static_cast<uint32_t>(hash);
+
+  // Calculate mask for bucket.
+  uint32_t mask[8];
+  setMask(key, mask);
+
+  for (int i = 0; i < 8; i++) {
+    bitset32[bucketIndex * 8 + i] |= mask[i];
+  }
+}
+
+bool Bloom::contains(uint64_t hash) {
+  uint32_t * const bitset32 = reinterpret_cast<uint32_t *>(bitset.get());
+  const uint32_t bucketIndex = static_cast<uint32_t>((hash >> 32)
+      & (num_bytes / BYTES_PER_FILTER_BLOCK - 1));
+  uint32_t key = static_cast<uint32_t>(hash);
+
+  // Calculate mask for bucket.
+  uint32_t mask[8];
+  setMask(key, mask);
+
+  for (int i = 0; i < 8; ++i) {
+    if (0 == (bitset32[8 * bucketIndex + i] & mask[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool Bloom::find(uint64_t hash) {
+  return contains(hash);
+}
+
+void Bloom::insert(uint64_t hash) {
+  addElement(hash);
+}
+
+uint64_t Bloom::hash(int value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(&value), sizeof(int), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+uint64_t Bloom::hash(int64_t value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void*>(&value), sizeof(int64_t), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+uint64_t Bloom::hash(float value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(&value), sizeof(float), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+
+uint64_t Bloom::hash(double value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(&value), sizeof(double), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+
+uint64_t Bloom::hash(const Int96 *value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(const_cast<uint32_t*>(value->value)),
+    sizeof(value->value), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+uint64_t Bloom::hash(const ByteArray *value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(const_cast<uint8_t*>(value->ptr)),
+    value->len, DEFAULT_SEED, &out);
+  return out[0];
+}
+
+uint64_t Bloom::hash(const FLBA *value, uint32_t len) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(const_cast<uint8_t*>(value->ptr)),
+    len, DEFAULT_SEED, &out);
+  return out[0];
+}
+
+void Bloom::writeTo(const std::shared_ptr<OutputStream>& sink) {
+  sink->Write(reinterpret_cast<const uint8_t *>(&num_bytes), sizeof(uint32_t));
+  sink->Write(reinterpret_cast<const uint8_t *>(&hash_strategy), sizeof(uint32_t));
+  sink->Write(reinterpret_cast<const uint8_t *>(&algorithm), sizeof(uint32_t));
+  sink->Write(bitset.get(), num_bytes);
+}
+
+Bloom::~Bloom() {
```

Comment:
You can put this in .h with and use " = default;"

Reply:
done.